### PR TITLE
[TradingMode] add signals handling for all trading modes

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot-Commons/pull/264

### CODE REVIEW

File: octobot_trading/personal_data/orders/order_util.py [hunk 1]
```diff
@@ -354,3 +354,25 @@ def get_order_quantity_currency(exchange_manager, symbol, side):
         return quote if position.symbol_contract.is_inverse_contract() else base
     # always base in spot
     return base
+
+
+async def get_order_size_portfolio_percent(exchange_manager, order_amount, side, symbol):
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/order_util.py:359]: 👍

Comment (Herklos) [reply octobot_trading/personal_data/orders/order_util.py:359]: Can we add a test for it please?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 2]
```diff
@@ -119,6 +128,15 @@ def is_trading_signal_emitter(self) -> bool:
         except KeyError:
             return False
 
+    def get_trading_signal_identifier(self) -> str:
+        """
+        :return: True if the mode should be emitting trading signals
+        """
+        try:
+            return self.trading_config[common_constants.CONFIG_TRADING_SIGNALS_STRATEGY]
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:136]: 👍

File: octobot_trading/modes/abstract_trading_mode.py [hunk 3]
```diff
@@ -119,6 +128,15 @@ def is_trading_signal_emitter(self) -> bool:
         except KeyError:
             return False
 
+    def get_trading_signal_identifier(self) -> str:
+        """
+        :return: True if the mode should be emitting trading signals
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:133]: It's not returning a boolean right?

**Review by Herklos — COMMENTED** [review]
Nice feature!

File: octobot_trading/personal_data/orders/order_util.py (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/orders/order_util.py:359]: Yes !

File: octobot_trading/modes/abstract_trading_mode.py [hunk 4]
```diff
@@ -112,13 +120,29 @@ def get_supported_exchange_types(cls) -> list:
...
         try:
             return self.trading_config[common_constants.CONFIG_EMIT_TRADING_SIGNALS]
         except KeyError:
             return False
 
+    def should_emit_trading_signal(self) -> bool:
+        """
+        :return: True if the mode should be emitting trading signals according to configuration and trading environment
+        """
+        # TODO
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:134]: Is there still a TODO here?

File: octobot_trading/modes/abstract_trading_mode.py [hunk 5]
```diff
@@ -262,3 +286,68 @@ def get_required_candles_count(cls, tentacles_setup_config: tm_configuration.Ten
...
+    async def remote_signal_publisher(self, symbol: str):
+        if self.should_emit_trading_signal():
+            try:
+                async with signals.SignalPublisher.instance().remote_signal_bundle_builder(
+                    symbol,
+                    self.get_trading_signal_identifier(),
+                    self.TRADING_SIGNAL_TIMEOUT,
+                    signals.TradingSignalBundleBuilder
+                ) as signal_builder:
+                    yield signal_builder
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:300]: Very clean 👌

File: octobot_trading/modes/abstract_trading_mode.py [hunk 6]
```diff
@@ -262,3 +286,68 @@ def get_required_candles_count(cls, tentacles_setup_config: tm_configuration.Ten
...
+            except authentication.AuthenticationRequired as e:
+                self.logger.exception(e, True, f"Failed to send trading signals: {e}")
+        else:
+            yield None
+
+    async def create_order(self, order, loaded: bool = False, params: dict = None, pre_init_callback=None):
+        order_pf_percent = "0"
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py:307]: Should it be `f"0{script_keywords.QuantityType.PERCENT.value}"` ?

File: octobot_trading/personal_data/orders/order_util.py [hunk 7]
```diff
@@ -354,3 +354,25 @@ def get_order_quantity_currency(exchange_manager, symbol, side):
...
+                                 portfolio_type=commons_constants.PORTFOLIO_TOTAL)
+    if exchange_manager.is_future:
+        # TODO check inverse
+        if market_quantity is constants.ZERO:
+            return constants.ZERO
+        return min(order_amount / market_quantity, constants.ONE) * constants.ONE_HUNDRED
+    if side is enums.TradeOrderSide.SELL:
+        if current_symbol_holding is constants.ZERO:
+            return constants.ZERO
+        return min(order_amount / current_symbol_holding, constants.ONE) * constants.ONE_HUNDRED
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/order_util.py:373]: We can't use < 1% pf order size?

File: octobot_trading/signals/trading_signal_bundle_builder.py [hunk 8]
```diff
@@ -13,43 +13,39 @@
...
-        self.strategy = strategy
-        self.exchange = exchange
-        self.exchange_type = exchange_type
-        self.symbol = symbol
-        self.description = description
-        self.state = state
-        self.orders = orders
-        self.reset()
-
+class TradingSignalBundleBuilder(signals.SignalBundleBuilder):
```
Comment (Herklos) [inline octobot_trading/signals/trading_signal_bundle_builder.py:23]: 👍

File: tests/modes/test_abstract_trading_mode.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+                should_emit_trading_signal_mock.reset_mock()
+            emit_signal_bundle_mock.assert_not_called()
+        with mock.patch.object(trading_mode, "should_emit_trading_signal", mock.Mock(return_value=True)) \
+                as should_emit_trading_signal_mock:
+            async with trading_mode.remote_signal_publisher("BTC/USDT") as builder:
+                assert await trading_mode.create_order(buy_limit_order, loaded=False, params=None,
+                                                       pre_init_callback=None) \
+                       is buy_limit_order
+                assert not builder.is_empty()
+                assert should_emit_trading_signal_mock.call_count == 3
```
Comment (Herklos) [inline tests/modes/test_abstract_trading_mode.py:118]: 👍

File: tests/personal_data/orders/test_order_util.py [hunk 10]
```diff
@@ -447,3 +448,38 @@ async def test_ensure_orders_relevancy_with_positions(future_trader_simulator_wi
...
+    assert await personal_data.get_order_size_portfolio_percent(
+        exchange_manager_inst, decimal.Decimal("0.1"), enums.TradeOrderSide.BUY, "BTC/UDST"
+    ) == constants.ZERO
+    # 10 BTC in portfolio
+    assert await personal_data.get_order_size_portfolio_percent(
+        exchange_manager_inst, decimal.Decimal("1"), enums.TradeOrderSide.SELL, "BTC/UDST"
+    ) == decimal.Decimal("10")
+    assert await personal_data.get_order_size_portfolio_percent(
+        exchange_manager_inst, decimal.Decimal("10"), enums.TradeOrderSide.SELL, "BTC/UDST"
+    ) == decimal.Decimal("100")
```
Comment (Herklos) [inline tests/personal_data/orders/test_order_util.py:467]: 👍

File: octobot_trading/modes/abstract_trading_mode.py (hunk 6 above)
Comment (GuillaumeDSM) [reply octobot_trading/modes/abstract_trading_mode.py:307]: yes !

File: octobot_trading/personal_data/orders/order_util.py (hunk 7 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/orders/order_util.py:373]: this means we can't use more than 100% of portfolio, less than 1 is allowed

File: octobot_trading/signals/trading_signal_bundle_builder.py [hunk 11]
```diff
@@ -155,9 +171,39 @@ def _update_pending_orders(
             pass
         return False
 
-    def get_order_description_from_local_orders(self, order):
-        for index, order_description in enumerate(self.signals):
-            if order_description.content[trading_enums.TradingSignalOrdersAttrs.SHARED_SIGNAL_ORDER_ID.value] == \
-                    order.shared_signal_order_id:
-                return index, order_description.content
-        raise trading_errors.OrderDescriptionNotFoundError(f"order not found {order}")
+    def _link_bundled_chained_and_grouped_orders(self):
```
Comment (Herklos) [inline octobot_trading/signals/trading_signal_bundle_builder.py:174]: 👍

File: tests/signals/test_trading_signal_bundle_builder.py [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+    assert trading_signal_bundle_builder.signals[0].content[enums.TradingSignalOrdersAttrs.UPDATED_CURRENT_PRICE.value] == 1.0
+
+    # add new order (orders are based on shared_signal_order_id)
+    buy_limit_order.set_shared_signal_order_id("other_id")
+    trading_signal_bundle_builder.add_edited_order(buy_limit_order, updated_target_position="1%a")
+    assert len(trading_signal_bundle_builder.signals) == 2
+    assert trading_signal_bundle_builder.signals[0].content[enums.TradingSignalOrdersAttrs.UPDATED_CURRENT_PRICE.value] == 1.0
+    assert trading_signal_bundle_builder.signals[0].content[enums.TradingSignalOrdersAttrs.UPDATED_TARGET_POSITION.value] == None
+    assert trading_signal_bundle_builder.signals[1].content[enums.TradingSignalOrdersAttrs.UPDATED_CURRENT_PRICE.value] == 0.0
+    assert trading_signal_bundle_builder.signals[1].content[enums.TradingSignalOrdersAttrs.UPDATED_TARGET_POSITION.value] == "1%a"
```
Comment (Herklos) [inline tests/signals/test_trading_signal_bundle_builder.py:221]: What's this `a` at the end of `"1%a"`?

Comment (GuillaumeDSM) [reply tests/signals/test_trading_signal_bundle_builder.py:221]: % of available, it's from scripted lib

**Review by Herklos — APPROVED** [review]
:100:

**Pull request merged**